File: scripts/pptx-template/theme.py
"""DS token -> OOXML theme (clrScheme/fontScheme) writer (data-model.md §4).

Overrides the presentation's ThemePart so that DS-token-derived colors and
Noto Sans JP become the deck-wide default (Clarification #2 in spec.md): not
just placeholder fills, but the default color/font PowerPoint offers for any
shape a user draws later.
"""

from lxml import etree
from pptx.opc.constants import RELATIONSHIP_TYPE as RT
from pptx.oxml import parse_xml

A_NS = "http://schemas.openxmlformats.org/drawingml/2006/main"
NSMAP = {"a": A_NS}

# OOXML theme color slot -> DS token role (data-model.md §4 ThemeMapping).
THEME_SLOT_MAP = {
    "dk1": "text-primary",
    "lt1": "slide-bg",
    "dk2": "text-secondary",
    "lt2": "surface",
    "accent1": "accent",
    "accent2": "accent-strong",
    "accent3": "accent-weak",
    "accent4": "state-success",
    "accent5": "state-error",
    "accent6": "state-warning",
    "hlink": "accent",
    "folHlink": "accent",
}

FONT_ROLE = "font-sans"


def _qn(tag: str) -> str:
    prefix, local = tag.split(":")
    return f"{{{NSMAP[prefix]}}}{local}"


def _set_solid_color(slot_element: etree._Element, hex_color: str) -> None:
    """Replace a clrScheme slot's child color element with a fixed srgbClr."""
    hex_color = hex_color.lstrip("#").upper()
    for child in list(slot_element):
        slot_element.remove(child)
    srgb = etree.SubElement(slot_element, _qn("a:srgbClr"))
    srgb.set("val", hex_color)


def apply_theme(prs, token_values: dict) -> None:
    """Overwrite theme color/font scheme in-place on *prs* using *token_values*.

    `token_values` maps DS token role names (e.g. "accent", "text-primary",
    "font-sans") to resolved values: colors as "#RRGGBB" strings, fonts as
    typeface name strings.
    """
    theme_part = prs.slide_masters[0].part.part_related_by(RT.THEME)
    root = parse_xml(theme_part.blob)

    clr_scheme = root.find(f".//{_qn('a:clrScheme')}")
    for slot_name, token_role in THEME_SLOT_MAP.items():
        slot_element = clr_scheme.find(_qn(f"a:{slot_name}"))
        if slot_element is None:
            continue
        color_hex = token_values.get(token_role)
        if not color_hex:
            continue
        _set_solid_color(slot_element, color_hex)

    font_family = token_values.get(FONT_ROLE)
    if font_family:
        for font_group in ("majorFont", "minorFont"):
            latin = root.find(f".//{_qn(f'a:{font_group}')}/{_qn('a:latin')}")
            if latin is not None:
                latin.set("typeface", font_family)

    # ThemePart is a plain (non-XML-aware) opc.package.Part in python-pptx: its
    # `.blob` returns the raw bytes captured at load time, it does not
    # re-serialize from an `_element`. Write the edited XML back explicitly.
    theme_part._blob = etree.tostring(
        root, xml_declaration=True, encoding="UTF-8", standalone=True
    )

File: scripts/pptx/theme.mjs
#!/usr/bin/env node
/**
 * theme.mjs — 生成済み PPTX の配色スキーム(`ppt/theme/theme1.xml`)を DS へ差し替える。
 *
 * pptxgenjs は Office 既定スキーム(accent1=#4472C4 ほか)を固定出力するため、
 * PowerPoint 上で利用者が **後から挿入する**図形やグラフの既定色が Office 青になる。
 * ここで accent1 を DS プライマリ青、accent2–6 を DS カテゴリ色、text2/bg2 を
 * DS グレーへ差し替えることで、後挿入のオブジェクトも DS 配色に乗る。
 *
 * 配色値は手維持のマップではなく、抽出時に正本の CSS 変数から解決した値を受け取る
 * (feature 011 / FR-010 — `tokens.mjs` の手写経マップは廃止)。
 */

/** Office のスロット -> DS の役割。抽出した解決値から埋める。 */
export function themeMapFrom(tokens) {
  const cat = tokens.cat || [];
  return {
    accent1: tokens.accent,
    accent2: cat[1] || tokens.accent,
    accent3: cat[2] || tokens.accent,
    accent4: cat[3] || tokens.accent,
    accent5: cat[4] || tokens.accent,
    accent6: cat[5] || tokens.accent,
    dk2: tokens.textPrimary,
    lt2: tokens.surface,
    hlink: tokens.accent,
    folHlink: tokens.accentStrong || tokens.accent,
  };
}

/** theme1.xml の配色スロットを置き換える。未知のスロットは触らない。 */
export function applyDsTheme(themeXml, map) {
  let xml = themeXml;
  for (const [slot, hex] of Object.entries(map)) {
    if (!hex) continue;
    const re = new RegExp(`(<a:${slot}>\\s*<a:srgbClr val=")[0-9A-Fa-f]{6}("\\s*/>)`, 'g');
    xml = xml.replace(re, `$1${hex.toUpperCase()}$2`);
    // dk2/lt2 は sysClr で書かれる場合がある
    const reSys = new RegExp(`(<a:${slot}>\\s*)<a:sysClr[^/]*/>(\\s*</a:${slot}>)`, 'g');
    xml = xml.replace(reSys, `$1<a:srgbClr val="${hex.toUpperCase()}"/>$2`);
  }
  return xml;
}
